### PR TITLE
Basic CHAP auth support added

### PR DESCRIPTION
**Pull request by rkojedzinszky**

### CODE REVIEW

**Comment by travisghansen** [comment]
Thanks for the contribution! It's going to take me a little bit to dig through this and make sure it's compatible with existing deployments. Anything I should be aware of generally?

On a side note, I've started reimplementing both this project and the nfs variant as a `csi` driver. I've basically completed the majority of the work but was holding off a bit for some larger community related things to come together. I'm just going to proceed however before those things are resolved and provide migration stuff as necessary.

It is however written in nodejs instead of go as I'm basically useless in go :( if you have some interest is joining that effort let me know.

**Comment by rkojedzinszky** [comment]
As new fields are being parsed of a StorageClass, on existing deployments with default values it should not change anything. Basically, if one wants the iSCSI pv to have the relevant fields, she/he must set at least one of the two new booleans to `true`.

The csi project seems to be interesting, although that is totally a new topic to me. Also, I am not an expert in nodejs, neither in go at the moment, so at the moment I dont think that I can jump in. :(

**Comment by travisghansen** [comment]
Yeah fair enough. I've got several comments I'll add shortly.

File: cli/command.go [hunk 1]
```diff
@@ -9,12 +9,12 @@ import (
 
 	"github.com/golang/glog"
 	cli "github.com/jawher/mow.cli"
-	"github.com/kubernetes-incubator/external-storage/lib/controller"
```
Comment (travisghansen) [inline cli/command.go:12]: Should we remove this from `glide.yaml`?

File: cli/command.go [hunk 2]
```diff
@@ -9,12 +9,12 @@ import (
...
 	freenasProvisioner "github.com/travisghansen/freenas-iscsi-provisioner/provisioner"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/sig-storage-lib-external-provisioner/controller"
```
Comment (travisghansen) [inline cli/command.go:17]: Should we add this to `glide.yaml`?

File: deploy/freenas-iscsi-chap.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+kind: Secret
+apiVersion: v1
+type: kubernetes.io/iscsi-chap
+metadata:
+  namespace: kube-system
+  name: freenas-iscsi-chap
+stringData:
+  # Set the relevant auth credentials here
```
Comment (travisghansen) [inline deploy/freenas-iscsi-chap.yaml:8]: This is new to me, let's make a note/comment here about which `base64` values are required or not.

File: provisioner/provisioner.go [hunk 4]
```diff
@@ -8,11 +8,11 @@ import (
 	"time"
 
 	"github.com/golang/glog"
-	"github.com/kubernetes-incubator/external-storage/lib/controller"
 	"github.com/travisghansen/freenas-iscsi-provisioner/freenas"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/sig-storage-lib-external-provisioner/controller"
```
Comment (travisghansen) [inline provisioner/provisioner.go:15]: Same question about `glide.yaml` here.

File: provisioner/provisioner.go [hunk 5]
```diff
@@ -153,6 +166,16 @@ func (p *freenasProvisioner) GetConfig(storageClassName string) (*freenasProvisi
 		case "targetGroupPortalgroup":
 			targetGroupPortalgroup, _ = strconv.Atoi(v)
 
+			// Authentication options
```
Comment (travisghansen) [inline provisioner/provisioner.go]: This looks like the comment may have a tab in it? It's pushed over further than the rest of the code in the diff interface in github.

File: provisioner/provisioner.go [hunk 6]
```diff
@@ -554,7 +590,7 @@ func (p *freenasProvisioner) Provision(options controller.VolumeOptions) (*v1.Pe
 			},
 		},
 		Spec: v1.PersistentVolumeSpec{
-			PersistentVolumeReclaimPolicy: options.PersistentVolumeReclaimPolicy,
+			PersistentVolumeReclaimPolicy: *config.ReclaimPolicy,
```
Comment (travisghansen) [inline provisioner/provisioner.go:593]: This is much better than what I had, thanks for adding!

File: provisioner/provisioner.go [hunk 7]
```diff
@@ -563,16 +599,16 @@ func (p *freenasProvisioner) Provision(options controller.VolumeOptions) (*v1.Pe
...
+					TargetPortal:      config.ProvisionerTargetPortal,
+					Portals:           portals,
+					IQN:               iscsiConfig.Basename + ":" + iscsiName,
+					ISCSIInterface:    config.ProvisionerISCSIInterface,
+					Lun:               int32(*targetToExtent.Lunid),
+					ReadOnly:          extent.Ro,
+					FSType:            config.FSType,
+					DiscoveryCHAPAuth: config.DiscoveryCHAPAuth,
+					SessionCHAPAuth:   config.SessionCHAPAuth,
+					SecretRef:         config.AuthSecretRef,
```
Comment (travisghansen) [inline provisioner/provisioner.go:611]: I presume this value is `nil` in the scenario where everything is off? Did you happen to run through some tests with auth disabled?

**Review by travisghansen — COMMENTED** [review]
This all looks really good. Excited to have the support make it in there! Couple minor things to discuss and then we'll get it merged!

File: deploy/freenas-iscsi-chap.yaml (hunk 3 above)
Comment (rkojedzinszky) [reply deploy/freenas-iscsi-chap.yaml:8]: I would recommend the user to be familiar with iscsi pv settings, more info here: https://github.com/kubernetes/examples/tree/master/volumes/iscsi

File: provisioner/provisioner.go (hunk 5 above)
Comment (rkojedzinszky) [reply provisioner/provisioner.go]: fixed

File: provisioner/provisioner.go (hunk 7 above)
Comment (rkojedzinszky) [reply provisioner/provisioner.go:611]: I've done manual testing, and it worked with nil SecretRef. Basically, that is the default value for that field I presume.

File: cli/command.go (hunk 1 above)
Comment (rkojedzinszky) [reply cli/command.go:12]: I am very sorry, I am not familiar with glide, so basically, I dont know, what and how to do here. :(

File: deploy/freenas-iscsi-chap.yaml (hunk 3 above)
Comment (travisghansen) [reply deploy/freenas-iscsi-chap.yaml:8]: I see, wasn't familiar with stringData option generally. That's a gem. Let's throw that link somewhere in the project.

File: cli/command.go (hunk 1 above)
Comment (travisghansen) [reply cli/command.go:12]: Ok, we may need to merge this first to a feature branch then and massage that aspect before hitting master.

File: deploy/freenas-iscsi-chap.yaml (hunk 3 above)
Comment (rkojedzinszky) [reply deploy/freenas-iscsi-chap.yaml:8]: Oh, you mean the stringData notion. :) For secrets, you can use stringData as a shorthand, that will be base64 encoded. :)

Comment (travisghansen) [reply deploy/freenas-iscsi-chap.yaml:8]: Yeah, I had never seen that :)

**Comment by travisghansen** [comment]
Can you change the target branch of this to `chap` on my side? I'll pull it in and work with the glide stuff.

**Comment by travisghansen** [comment]
I pulled this into a branch locally and reverted the dependency change. All is well and pushed into `next` branch along with creating docker images with the `next` tag. I'll merge to master in the next day or two and tag as well.

**Comment by rkojedzinszky** [comment]
Sorry, I was a bit lost in the past week. Anyway, thank for reviewing and accepting the patch. Keep up the good work!

**Comment by travisghansen** [comment]
Thanks for the contribution! I reverted the dependency updates as they weren't needed and it was causing quite the headache to build etc. At this point all seems in order.

I really need to wrap up the csi driver I'm working on though to make this all a bit future proof and generally sunset this project.

**Pull request merged**